lib/Dialect/HW/CustomDirectiveImpl.cpp

```diff
@@ -0,0 +1,137 @@
+//===- CustomDirectiveImpl.cpp - Custom directive definitions -------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/HW/CustomDirectiveImpl.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+
+using namespace circt;
+using namespace circt::hw;
+
+ParseResult circt::parseInputPortList(
+    OpAsmParser &parser, SmallVector<OpAsmParser::UnresolvedOperand, 4> &inputs,
+    SmallVector<Type, 1> &inputTypes, ArrayAttr &inputNames) {
+
+  SmallVector<Attribute> argNames;
+  auto parseInputPort = [&]() -> ParseResult {
+    std::string portName;
+    if (parser.parseKeywordOrString(&portName))
+      return failure();
+    argNames.push_back(StringAttr::get(parser.getContext(), portName));
+    inputs.push_back({});
+    inputTypes.push_back({});
+    return failure(parser.parseColon() || parser.parseOperand(inputs.back()) ||
+                   parser.parseColon() || parser.parseType(inputTypes.back()));
+  };
+  llvm::SMLoc inputsOperandsLoc;
+  if (parser.getCurrentLocation(&inputsOperandsLoc) ||
+      parser.parseCommaSeparatedList(OpAsmParser::Delimiter::Paren,
+                                     parseInputPort))
+    return failure();
+
+  inputNames = ArrayAttr::get(parser.getContext(), argNames);
+
+  return success();
+}
+
+void circt::printInputPortList(OpAsmPrinter &p, Operation *op,
+                               OperandRange inputs, TypeRange inputTypes,
+                               ArrayAttr inputNames) {
+  p << "(";
+  llvm::interleaveComma(llvm::zip(inputs, inputNames), p, [&](auto input) {
+    Value val = std::get<0>(input);
+    p.printKeywordOrString(
+        std::get<1>(input).template cast<StringAttr>().getValue());
+    p << ": " << val << ": " << val.getType();
+  });
```

Comment:
I think the `.template` is fine, but you may want to check with MSVC. I know that using an explicit type instead of `auto input` is safe. Ditto below.

Reply:
Let's play it safe then.